<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/06_labeling_states.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB06 — Derivação de Estados e Formulação Supervisionada

## 1. Contexto

A partir da base temporal enriquecida produzida no NB05 e dos episódios críticos detectados no NB04, este notebook formaliza a estrutura supervisionada do problema em nível de janelas temporais de 5 minutos.

Nesta etapa, a série deixa de ser apenas uma sequência de atributos e episódios e passa a ser organizada em:

1. estados operacionais por janela;
2. relações temporais entre janelas consecutivas;
3. bases supervisionadas voltadas à antecipação de entrada em estado crítico.

A versão anterior utilizava `K = 12` janelas para definir o estado `BEFORE` e `H = 12` janelas como horizonte de antecipação. Como os dois parâmetros eram iguais, a classe positiva ficava fortemente associada ao estado `BEFORE`, o que poderia gerar circularidade metodológica.

Esta versão separa explicitamente:

1. o parâmetro `K`, usado para rotular estados operacionais;
2. o parâmetro `H`, usado para definir o alvo supervisionado de antecipação.

Essa separação permite avaliar se o alvo positivo continua sendo apenas uma consequência direta da definição do estado `BEFORE`, ou se passa a representar uma tarefa supervisionada mais informativa.

## 2. Objetivo

O objetivo principal deste notebook é derivar estados operacionais e construir bases supervisionadas para antecipação de entrada em estado crítico.

Especificamente, este notebook busca:

1. Ler a base de features `window_5min_features.parquet`, produzida no NB05;
2. Ler os episódios críticos oficiais `episodes_detected.parquet`, produzidos no NB04;
3. Ajustar os episódios à faixa efetivamente disponível na base de features;
4. Rotular cada janela temporal como `NORMAL`, `BEFORE`, `DURING` ou `AFTER`;
5. Aplicar precedência entre estados para resolver sobreposições;
6. Derivar transições de um passo entre estados consecutivos;
7. Construir bases supervisionadas binárias de antecipação;
8. Comparar diferentes combinações entre `K` e `H`;
9. Diagnosticar a equivalência entre estado `BEFORE` e alvo positivo;
10. Definir um cenário principal revisado;
11. Persistir artefatos canônicos e artefatos por cenário;
12. Gerar summaries JSON com rastreabilidade da formulação.

## 3. Papel no Pipeline

O NB06 representa a passagem entre:

1. a segmentação estatística de episódios críticos;
2. a engenharia de atributos temporais;
3. a formulação supervisionada utilizada pelos modelos posteriores.

Seu papel é transformar a série temporal enriquecida em três tipos de artefatos:

1. uma base rotulada por estados;
2. uma base supervisionada principal de antecipação;
3. bases de sensibilidade para avaliar a dependência dos resultados em relação a `K` e `H`.

Assim, o NB06 consolida a transição entre a descrição temporal do comportamento da série e a tarefa preditiva efetivamente avaliada no restante do pipeline.

## 4. Definição dos Estados

Cada janela é rotulada em uma das seguintes categorias:

| Estado | Definição |
|---|---|
| `DURING` | Janela pertencente a um episódio crítico efetivamente disponível na base de features |
| `BEFORE` | Janela situada até `K` passos antes do início de um episódio |
| `AFTER` | Janela situada até `K` passos após o fim de um episódio |
| `NORMAL` | Janela fora das categorias anteriores |

A precedência adotada para evitar conflitos é:

1. `DURING`;
2. `BEFORE`;
3. `AFTER`;
4. `NORMAL`.

Essa precedência garante unicidade da rotulagem quando há sobreposição entre regiões associadas a episódios distintos.

## 5. Ajuste entre Episódios Detectados e Base de Features

Os episódios detectados no NB04 referem-se à série scored completa. Já a base do NB05 pode não conter todas as janelas originais, devido ao trim inicial necessário para criação dos lags.

Por isso, esta etapa trabalha com episódios efetivos:

1. episódios integralmente contidos na faixa de buckets da base de features; ou
2. episódios recortados pela interseção com essa faixa, quando necessário.

Esse procedimento evita inconsistências entre:

1. o total de janelas críticas disponível na base de atributos;
2. a marcação `DURING` utilizada na formulação supervisionada;
3. os episódios originalmente detectados sobre a série completa do NB04.

## 6. Formulação Supervisionada

A formulação supervisionada construída nesta etapa é binária.

Para cada janela elegível no estado atual `NORMAL` ou `BEFORE`, define-se um horizonte futuro de `H` janelas.

O alvo supervisionado é definido como:

| Valor | Definição |
|---|---|
| `1` | Haverá pelo menos uma janela em estado `DURING` dentro do horizonte futuro |
| `0` | Não haverá janela `DURING` dentro do horizonte futuro |

Diferentemente da versão anterior, a classe `0` não exige que todo o horizonte futuro permaneça `NORMAL`. Isso evita excluir artificialmente janelas que ainda não entram em criticidade dentro de `H`, mas que podem estar próximas de uma região pré-crítica.

## 7. Cenários Avaliados

São avaliados os seguintes cenários:

| Cenário | K | H | Interpretação |
|---|---:|---:|---|
| `legacy_K12_H12` | 12 | 12 | Cenário legado, com janela de estado e horizonte preditivo iguais |
| `sensitivity_K12_H06` | 12 | 6 | Sensibilidade com horizonte de 30 minutos |
| `sensitivity_K12_H03` | 12 | 3 | Sensibilidade com horizonte de 15 minutos |
| `primary_K24_H12` | 24 | 12 | Cenário principal revisado |
| `sensitivity_K24_H06` | 24 | 6 | Sensibilidade com K ampliado e horizonte de 30 minutos |
| `sensitivity_K24_H03` | 24 | 3 | Sensibilidade com K ampliado e horizonte de 15 minutos |

O cenário `primary_K24_H12` preserva o horizonte preditivo de 60 minutos, mas usa uma janela de pré-criticidade mais ampla, de 120 minutos, para evitar equivalência automática entre `BEFORE` e classe positiva.

## 8. Diagnóstico de Circularidade

Para cada cenário, este notebook calcula:

1. distribuição dos estados;
2. distribuição do alvo supervisionado;
3. tabela cruzada `state × transition_target`;
4. quantidade de janelas `NORMAL` com alvo 1;
5. quantidade de janelas `BEFORE` com alvo 0;
6. quantidade de janelas `BEFORE` com alvo 1;
7. verificação se `transition_target = 1` é equivalente ao estado `BEFORE`;
8. verificação se todos os positivos pertencem ao estado `BEFORE`;
9. verificação se todas as janelas `BEFORE` são positivas.

Esse diagnóstico é central para avaliar se a tarefa supervisionada está realmente formulada como antecipação, ou apenas como identificação direta do estado `BEFORE`.

## 9. Artefatos Utilizados e Gerados

Este notebook utiliza como entrada:

| Artefato | Origem | Finalidade |
|---|---|---|
| `window_5min_features.parquet` | NB05 | Base temporal enriquecida com atributos |
| `episodes_detected.parquet` | NB04 | Episódios críticos oficiais |

O notebook gera os artefatos principais:

| Artefato | Finalidade |
|---|---|
| `window_5min_labeled.parquet` | Base rotulada do cenário principal |
| `anticipation_supervised_dataset.parquet` | Base supervisionada principal |
| `transition_dataset.parquet` | Alias legado para compatibilidade |
| `06_labeling_states_summary.json` | Summary principal |
| `06_horizon_sensitivity_summary.json` | Summary de sensibilidade |

Também são gerados artefatos por cenário, com nomes explícitos contendo `K` e `H`, permitindo auditoria e comparação posterior.

## 10. Relação com as Próximas Etapas

Este notebook alimenta diretamente:

1. NB08 — modelagem supervisionada e validação temporal;
2. NB09 — consolidação experimental e geração de figuras finais.

A qualidade da formulação supervisionada construída aqui impacta diretamente a validade das etapas de modelagem e avaliação posteriores.

## 11. Observações Metodológicas

Esta etapa preserva causalidade temporal nos atributos, pois as features vêm do NB05 e já foram construídas sem uso de informação futura.

O alvo supervisionado, por definição, usa o futuro apenas para construção retrospectiva dos rótulos, o que é compatível com aprendizagem supervisionada.

A principal mudança desta versão é metodológica: separar a janela usada para descrever o estado `BEFORE` do horizonte usado para definir o alvo de antecipação. Essa separação reduz o risco de circularidade entre rotulagem e predição.

Mesmo no cenário principal revisado, os exemplos positivos continuam dentro do estado `BEFORE`. Portanto, a tarefa deve ser interpretada como antecipação dentro de uma região pré-crítica ampliada, e não como predição totalmente independente da definição dos estados operacionais.

In [ ]:
# ============================================================
# NB06 — Derivação de Estados e Formulação Supervisionada
# Pipeline PPCOMP_DM
#
# Escopo:
# - Ler a base de features produzida no NB05
# - Ler os episódios críticos oficiais detectados no NB04
# - Ajustar episódios à faixa efetivamente disponível nas features
# - Rotular janelas como NORMAL, BEFORE, DURING ou AFTER
# - Construir bases supervisionadas para diferentes combinações de K e H
# - Diagnosticar a equivalência entre estado BEFORE e alvo positivo
# - Definir o cenário principal revisado primary_K24_H12
# - Persistir artefatos canônicos, artefatos por cenário e summaries JSON
#
# Importante:
# - K define a largura dos estados BEFORE e AFTER
# - H define o horizonte futuro do alvo supervisionado
# - O cenário legado K12_H12 é mantido apenas para comparação
# - O cenário principal revisado é K24_H12
# ============================================================

from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd
from IPython.display import display

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# ─────────────────────────────────────────────────────────────
# BLOCO 0 — Bootstrap
# ─────────────────────────────────────────────────────────────

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

# Garante o repositório local no Google Drive, clonando apenas se necessário.
if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull)...")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso: não foi possível atualizar via git pull:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs

ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH =", REPORTS_PATH)

def log(msg: str) -> None:
    """Imprime mensagens padronizadas desta etapa do pipeline."""
    print(f"[NB06_labeling_states] {msg}")

# ─────────────────────────────────────────────────────────────
# BLOCO 1 — Parâmetros
# ─────────────────────────────────────────────────────────────

WINDOW_MINUTES = 5

STATE_NORMAL = "NORMAL"
STATE_BEFORE = "BEFORE"
STATE_DURING = "DURING"
STATE_AFTER = "AFTER"

# Precedência usada para resolver sobreposições entre regiões de episódios.
STATE_PRECEDENCE = {
    STATE_NORMAL: 0,
    STATE_AFTER: 1,
    STATE_BEFORE: 2,
    STATE_DURING: 3,
}

# Cenário principal revisado:
# K=24 amplia a região BEFORE/AFTER para 120 minutos.
# H=12 mantém o horizonte supervisionado de 60 minutos.
PRIMARY_K = 24
PRIMARY_H = 12
PRIMARY_SCENARIO_NAME = f"primary_K{PRIMARY_K:02d}_H{PRIMARY_H:02d}"

# Cenários de sensibilidade usados para diagnosticar circularidade e horizonte.
SCENARIOS = [
    {"name": "legacy_K12_H12", "K": 12, "H": 12},
    {"name": "sensitivity_K12_H06", "K": 12, "H": 6},
    {"name": "sensitivity_K12_H03", "K": 12, "H": 3},
    {"name": "primary_K24_H12", "K": 24, "H": 12},
    {"name": "sensitivity_K24_H06", "K": 24, "H": 6},
    {"name": "sensitivity_K24_H03", "K": 24, "H": 3},
]

FEATURES_FILE = FEATURES_PATH / "window_5min_features.parquet"
EPISODES_FILE = FEATURES_PATH / "episodes_detected.parquet"

OUT_LABELED_PRIMARY = FEATURES_PATH / "window_5min_labeled.parquet"
OUT_SUPERVISED_PRIMARY = FEATURES_PATH / "anticipation_supervised_dataset.parquet"

# Alias legado mantido para compatibilidade com notebooks posteriores.
OUT_TRANSITION_LEGACY_ALIAS = FEATURES_PATH / "transition_dataset.parquet"

SUMMARY_FILE = REPORTS_PATH / "06_labeling_states_summary.json"
SENSITIVITY_SUMMARY_FILE = REPORTS_PATH / "06_horizon_sensitivity_summary.json"

# ─────────────────────────────────────────────────────────────
# BLOCO 2 — Leitura dos artefatos
# ─────────────────────────────────────────────────────────────

assert FEATURES_FILE.exists(), f"Arquivo não encontrado: {FEATURES_FILE}"
assert EPISODES_FILE.exists(), f"Arquivo não encontrado: {EPISODES_FILE}"

df_features = pd.read_parquet(FEATURES_FILE)
df_episodes = pd.read_parquet(EPISODES_FILE)

required_feature_cols = {"bucket_id", "bucket_start_us", "fail_rate", "is_critical"}
missing_features = required_feature_cols - set(df_features.columns)

assert not missing_features, (
    f"Colunas ausentes na base de features: {missing_features}"
)

required_episode_cols = {"episode_id", "start_bucket", "end_bucket"}
missing_episodes = required_episode_cols - set(df_episodes.columns)

assert not missing_episodes, (
    f"Colunas ausentes nos episódios: {missing_episodes}"
)

df_features = df_features.sort_values("bucket_id").reset_index(drop=True)
df_episodes = df_episodes.sort_values(
    ["start_bucket", "end_bucket"]
).reset_index(drop=True)

assert df_features["bucket_id"].is_monotonic_increasing, (
    "bucket_id não está ordenado na base de features."
)

assert df_features["bucket_id"].is_unique, (
    "bucket_id duplicado na base de features."
)

bucket_diff = df_features["bucket_id"].diff().dropna()
n_gaps = int((bucket_diff > 1).sum())

assert n_gaps == 0, (
    "A base de features deveria estar contínua em bucket_id após o trim inicial."
)

bucket_min = int(df_features["bucket_id"].min())
bucket_max = int(df_features["bucket_id"].max())

log(f"Features carregadas: shape={df_features.shape}")
log(f"Episódios carregados: shape={df_episodes.shape}")
log(f"Faixa de bucket_id nas features: {bucket_min}..{bucket_max}")

# ─────────────────────────────────────────────────────────────
# BLOCO 3 — Funções auxiliares
# ─────────────────────────────────────────────────────────────

def assign_state_with_precedence(
    state_array: np.ndarray,
    rank_array: np.ndarray,
    mask: np.ndarray,
    new_state: str,
) -> None:
    """
    Atribui um novo estado somente quando sua precedência é maior
    que a precedência já registrada para a janela.
    """
    new_rank = STATE_PRECEDENCE[new_state]
    update_mask = mask & (new_rank > rank_array)

    state_array[update_mask] = new_state
    rank_array[update_mask] = new_rank


def build_effective_episodes(
    episodes_df: pd.DataFrame,
    feature_bucket_min: int,
    feature_bucket_max: int,
) -> pd.DataFrame:
    """
    Ajusta os episódios detectados no NB04 à faixa de buckets disponível
    na base de features produzida no NB05.
    """
    rows = []

    for _, row in episodes_df.iterrows():
        original_start = int(row["start_bucket"])
        original_end = int(row["end_bucket"])

        effective_start = max(original_start, feature_bucket_min)
        effective_end = min(original_end, feature_bucket_max)

        if effective_start <= effective_end:
            new_row = row.to_dict()

            new_row["original_start_bucket"] = original_start
            new_row["original_end_bucket"] = original_end
            new_row["effective_start_bucket"] = effective_start
            new_row["effective_end_bucket"] = effective_end
            new_row["was_clipped_by_feature_range"] = (
                effective_start != original_start
                or effective_end != original_end
            )
            new_row["effective_duration_windows"] = int(
                effective_end - effective_start + 1
            )

            rows.append(new_row)

    effective_df = pd.DataFrame(rows)

    if len(effective_df) > 0:
        effective_df = effective_df.sort_values(
            ["effective_start_bucket", "effective_end_bucket"]
        ).reset_index(drop=True)

    return effective_df


def label_states_for_k(
    features_df: pd.DataFrame,
    effective_episodes_df: pd.DataFrame,
    k_before: int,
    k_after: int,
) -> pd.DataFrame:
    """
    Rotula cada bucket como NORMAL, BEFORE, DURING ou AFTER
    para uma janela K específica.
    """
    df = features_df.copy().sort_values("bucket_id").reset_index(drop=True)

    bucket_values = df["bucket_id"].to_numpy()

    state_array = np.array([STATE_NORMAL] * len(df), dtype=object)
    rank_array = np.zeros(len(df), dtype=np.int8)

    for _, ep in effective_episodes_df.iterrows():
        start_b = int(ep["effective_start_bucket"])
        end_b = int(ep["effective_end_bucket"])

        before_start = max(bucket_min, start_b - k_before)
        before_end = start_b - 1

        after_start = end_b + 1
        after_end = min(bucket_max, end_b + k_after)

        if after_start <= after_end:
            mask_after = (
                (bucket_values >= after_start)
                & (bucket_values <= after_end)
            )

            assign_state_with_precedence(
                state_array=state_array,
                rank_array=rank_array,
                mask=mask_after,
                new_state=STATE_AFTER,
            )

        if before_start <= before_end:
            mask_before = (
                (bucket_values >= before_start)
                & (bucket_values <= before_end)
            )

            assign_state_with_precedence(
                state_array=state_array,
                rank_array=rank_array,
                mask=mask_before,
                new_state=STATE_BEFORE,
            )

        mask_during = (
            (bucket_values >= start_b)
            & (bucket_values <= end_b)
        )

        assign_state_with_precedence(
            state_array=state_array,
            rank_array=rank_array,
            mask=mask_during,
            new_state=STATE_DURING,
        )

    df["state"] = state_array
    df["state_rank"] = rank_array
    df["state_window_k"] = int(k_before)

    # Transição de um passo, usada como diagnóstico estrutural.
    df["next_bucket_id"] = df["bucket_id"].shift(-1)
    df["next_state"] = df["state"].shift(-1)
    df["is_last_window"] = df["next_bucket_id"].isna()

    df["has_valid_next"] = (
        (~df["is_last_window"])
        & ((df["next_bucket_id"] - df["bucket_id"]) == 1)
    )

    df["transition"] = np.where(
        df["has_valid_next"],
        df["state"] + "->" + df["next_state"],
        pd.NA,
    )

    return df


def add_future_horizon_fields(
    labeled_df: pd.DataFrame,
    horizon_windows: int,
) -> pd.DataFrame:
    """
    Para cada janela, avalia os estados futuros dentro de H janelas.
    Esses campos são usados para construir o alvo supervisionado.
    """
    df = labeled_df.copy().sort_values("bucket_id").reset_index(drop=True)

    state_by_bucket = dict(zip(df["bucket_id"], df["state"]))

    future_valid_flags = []
    future_has_during_flags = []
    future_all_normal_flags = []
    future_path_states = []
    future_first_during_offset = []

    for b in df["bucket_id"].tolist():
        future_states = []
        valid_horizon = True

        for h in range(1, horizon_windows + 1):
            next_b = b + h
            s = state_by_bucket.get(next_b, None)

            if s is None:
                valid_horizon = False
                break

            future_states.append(s)

        if not valid_horizon:
            future_valid_flags.append(False)
            future_has_during_flags.append(False)
            future_all_normal_flags.append(False)
            future_path_states.append(pd.NA)
            future_first_during_offset.append(pd.NA)
            continue

        has_during = any(s == STATE_DURING for s in future_states)
        all_normal = all(s == STATE_NORMAL for s in future_states)

        first_during = pd.NA

        if has_during:
            for idx, s in enumerate(future_states, start=1):
                if s == STATE_DURING:
                    first_during = idx
                    break

        future_valid_flags.append(True)
        future_has_during_flags.append(has_during)
        future_all_normal_flags.append(all_normal)
        future_path_states.append(" | ".join(future_states))
        future_first_during_offset.append(first_during)

    df["future_horizon_valid"] = future_valid_flags
    df["future_has_during"] = future_has_during_flags
    df["future_all_normal"] = future_all_normal_flags
    df["future_path_states"] = future_path_states
    df["future_first_during_offset"] = future_first_during_offset
    df["target_horizon_windows"] = int(horizon_windows)
    df["prediction_horizon_minutes"] = int(horizon_windows * WINDOW_MINUTES)

    return df


def build_supervised_dataset(
    labeled_df: pd.DataFrame,
    horizon_windows: int,
    scenario_name: str,
) -> pd.DataFrame:
    """
    Constrói a base supervisionada binária para antecipação de DURING
    dentro de H janelas futuras.
    """
    df_h = add_future_horizon_fields(
        labeled_df=labeled_df,
        horizon_windows=horizon_windows,
    )

    eligible_mask = (
        df_h["state"].isin([STATE_NORMAL, STATE_BEFORE])
        & df_h["future_horizon_valid"]
    )

    supervised_df = df_h.loc[eligible_mask].copy()

    supervised_df["transition_target"] = np.where(
        supervised_df["future_has_during"],
        1,
        0,
    ).astype("int64")

    supervised_df["transition_label"] = np.where(
        supervised_df["transition_target"] == 1,
        f"TO_{STATE_DURING}_WITHIN_{horizon_windows}_WINDOWS",
        f"NO_{STATE_DURING}_WITHIN_{horizon_windows}_WINDOWS",
    )

    supervised_df["scenario_name"] = scenario_name
    supervised_df["target_definition"] = (
        "1 if DURING occurs within H future windows; 0 otherwise"
    )

    return supervised_df


def value_counts_dict(series: pd.Series) -> dict:
    """Converte value_counts em dicionário serializável em JSON."""
    return {
        str(k): int(v)
        for k, v in series.value_counts().sort_index().to_dict().items()
    }


def crosstab_to_nested_dict(ct: pd.DataFrame) -> dict:
    """Converte uma tabela cruzada em dicionário aninhado para JSON."""
    result = {}

    for idx in ct.index:
        result[str(idx)] = {
            str(col): int(ct.loc[idx, col])
            for col in ct.columns
        }

    return result


def summarize_scenario(
    scenario_name: str,
    k_before: int,
    horizon_windows: int,
    labeled_df: pd.DataFrame,
    supervised_df: pd.DataFrame,
    effective_episodes_df: pd.DataFrame,
) -> dict:
    """
    Resume a distribuição de estados, alvos e diagnósticos de circularidade
    para uma combinação específica de K e H.
    """
    state_counts = value_counts_dict(labeled_df["state"])
    target_counts = value_counts_dict(supervised_df["transition_target"])

    state_target_ct = pd.crosstab(
        supervised_df["state"],
        supervised_df["transition_target"],
        dropna=False,
    )

    # Garante colunas 0 e 1 mesmo quando uma classe estiver ausente.
    for col in [0, 1]:
        if col not in state_target_ct.columns:
            state_target_ct[col] = 0

    state_target_ct = state_target_ct[[0, 1]]

    normal_target_1 = 0
    before_target_0 = 0
    before_target_1 = 0

    if STATE_NORMAL in state_target_ct.index:
        normal_target_1 = int(state_target_ct.loc[STATE_NORMAL, 1])

    if STATE_BEFORE in state_target_ct.index:
        before_target_0 = int(state_target_ct.loc[STATE_BEFORE, 0])
        before_target_1 = int(state_target_ct.loc[STATE_BEFORE, 1])

    before_indicator = (supervised_df["state"] == STATE_BEFORE).astype("int64")
    target_values = supervised_df["transition_target"].astype("int64")

    if len(supervised_df) > 0:
        target_equals_before_state = bool(
            (before_indicator.values == target_values.values).all()
        )

        all_positive_are_before = bool(
            (
                supervised_df.loc[
                    supervised_df["transition_target"] == 1,
                    "state",
                ] == STATE_BEFORE
            ).all()
        )

        all_before_are_positive = bool(
            (
                supervised_df.loc[
                    supervised_df["state"] == STATE_BEFORE,
                    "transition_target",
                ] == 1
            ).all()
        )
    else:
        target_equals_before_state = False
        all_positive_are_before = False
        all_before_are_positive = False

    summary = {
        "scenario_name": scenario_name,
        "k_before": int(k_before),
        "k_after": int(k_before),
        "horizon_windows": int(horizon_windows),
        "horizon_minutes": int(horizon_windows * WINDOW_MINUTES),
        "features_rows": int(len(labeled_df)),
        "effective_episodes": int(len(effective_episodes_df)),
        "effective_during_windows": int(
            (labeled_df["state"] == STATE_DURING).sum()
        ),
        "state_distribution": state_counts,
        "supervised_rows": int(len(supervised_df)),
        "target_distribution": target_counts,
        "state_target_crosstab": crosstab_to_nested_dict(state_target_ct),
        "normal_target_1": int(normal_target_1),
        "before_target_0": int(before_target_0),
        "before_target_1": int(before_target_1),
        "target_equals_before_state": target_equals_before_state,
        "all_positive_are_before": all_positive_are_before,
        "all_before_are_positive": all_before_are_positive,
    }

    return summary


def save_json(path: Path, data: dict) -> None:
    """Salva dicionário em JSON UTF-8 indentado."""
    path.write_text(
        json.dumps(data, indent=2, ensure_ascii=False),
        encoding="utf-8",
    )

# ─────────────────────────────────────────────────────────────
# BLOCO 4 — Episódios efetivos
# ─────────────────────────────────────────────────────────────

effective_episodes = build_effective_episodes(
    episodes_df=df_episodes,
    feature_bucket_min=bucket_min,
    feature_bucket_max=bucket_max,
)

log(f"Episódios originais: {len(df_episodes)}")
log(f"Episódios efetivos na base de features: {len(effective_episodes)}")

if len(effective_episodes) == 0:
    raise RuntimeError(
        "Nenhum episódio efetivo encontrado na faixa da base de features."
    )

expected_during_effective = int(
    effective_episodes["effective_duration_windows"].sum()
)

log(f"Janelas DURING efetivas esperadas: {expected_during_effective}")

# ─────────────────────────────────────────────────────────────
# BLOCO 5 — Execução dos cenários
# ─────────────────────────────────────────────────────────────

scenario_summaries = []
scenario_artifacts = {}
labeled_by_k = {}

for scenario in SCENARIOS:
    scenario_name = scenario["name"]
    k = int(scenario["K"])
    h = int(scenario["H"])

    log(f"Executando cenário: {scenario_name} (K={k}, H={h})")

    # A rotulagem depende apenas de K; por isso é reaproveitada entre cenários
    # que compartilham o mesmo valor de K.
    if k not in labeled_by_k:
        labeled_df_k = label_states_for_k(
            features_df=df_features,
            effective_episodes_df=effective_episodes,
            k_before=k,
            k_after=k,
        )

        actual_during = int((labeled_df_k["state"] == STATE_DURING).sum())

        assert actual_during == expected_during_effective, (
            f"Inconsistência no total de janelas DURING para K={k}: "
            f"esperado={expected_during_effective}, observado={actual_during}"
        )

        labeled_by_k[k] = labeled_df_k

        labeled_k_file = FEATURES_PATH / f"window_5min_labeled_K{k:02d}.parquet"

        labeled_df_k.to_parquet(
            labeled_k_file,
            compression="snappy",
            index=False,
        )

        scenario_artifacts[f"labeled_K{k:02d}"] = str(labeled_k_file)

    labeled_df = labeled_by_k[k]

    supervised_df = build_supervised_dataset(
        labeled_df=labeled_df,
        horizon_windows=h,
        scenario_name=scenario_name,
    )

    supervised_file = (
        FEATURES_PATH / f"anticipation_supervised_dataset_K{k:02d}_H{h:02d}.parquet"
    )

    supervised_df.to_parquet(
        supervised_file,
        compression="snappy",
        index=False,
    )

    scenario_artifacts[scenario_name] = str(supervised_file)

    scenario_summary = summarize_scenario(
        scenario_name=scenario_name,
        k_before=k,
        horizon_windows=h,
        labeled_df=labeled_df,
        supervised_df=supervised_df,
        effective_episodes_df=effective_episodes,
    )

    scenario_summary["output_supervised_file"] = str(supervised_file)
    scenario_summaries.append(scenario_summary)

    print("\n" + "=" * 80)
    print(f"Cenário: {scenario_name}")
    print(f"K={k} | H={h}")

    print("Distribuição de estados:")
    display(
        pd.Series(
            scenario_summary["state_distribution"],
            name="count",
        ).to_frame()
    )

    print("Distribuição do alvo:")
    display(
        pd.Series(
            scenario_summary["target_distribution"],
            name="count",
        ).to_frame()
    )

    print("Tabela state × target:")
    display(
        pd.crosstab(
            supervised_df["state"],
            supervised_df["transition_target"],
            dropna=False,
        )
    )

    print("Diagnóstico de equivalência:")
    print("target_equals_before_state:", scenario_summary["target_equals_before_state"])
    print("all_positive_are_before:", scenario_summary["all_positive_are_before"])
    print("all_before_are_positive:", scenario_summary["all_before_are_positive"])
    print("NORMAL com target=1:", scenario_summary["normal_target_1"])
    print("BEFORE com target=0:", scenario_summary["before_target_0"])
    print("BEFORE com target=1:", scenario_summary["before_target_1"])

# ─────────────────────────────────────────────────────────────
# BLOCO 6 — Definição do cenário principal
# ─────────────────────────────────────────────────────────────

primary_summary_matches = [
    s for s in scenario_summaries
    if s["scenario_name"] == PRIMARY_SCENARIO_NAME
]

assert len(primary_summary_matches) == 1, (
    f"Cenário principal {PRIMARY_SCENARIO_NAME} não encontrado."
)

primary_summary = primary_summary_matches[0]

primary_k = int(primary_summary["k_before"])
primary_h = int(primary_summary["horizon_windows"])

df_labeled_primary = labeled_by_k[primary_k].copy()

df_supervised_primary = pd.read_parquet(
    FEATURES_PATH / f"anticipation_supervised_dataset_K{primary_k:02d}_H{primary_h:02d}.parquet"
)

# Para manter compatibilidade com os notebooks seguintes, os arquivos canônicos
# passam a representar o cenário principal revisado.
df_labeled_primary = add_future_horizon_fields(
    labeled_df=df_labeled_primary,
    horizon_windows=primary_h,
)

df_labeled_primary.to_parquet(
    OUT_LABELED_PRIMARY,
    compression="snappy",
    index=False,
)

df_supervised_primary.to_parquet(
    OUT_SUPERVISED_PRIMARY,
    compression="snappy",
    index=False,
)

# Alias legado para compatibilidade com notebooks posteriores.
df_supervised_primary.to_parquet(
    OUT_TRANSITION_LEGACY_ALIAS,
    compression="snappy",
    index=False,
)

log(f"Base rotulada principal salva: {OUT_LABELED_PRIMARY}")
log(f"Base supervisionada principal salva: {OUT_SUPERVISED_PRIMARY}")
log(f"Alias legado salvo: {OUT_TRANSITION_LEGACY_ALIAS}")

# ─────────────────────────────────────────────────────────────
# BLOCO 7 — Tabelas consolidadas
# ─────────────────────────────────────────────────────────────

summary_table_rows = []

for s in scenario_summaries:
    target_0 = int(s["target_distribution"].get("0", 0))
    target_1 = int(s["target_distribution"].get("1", 0))
    total = int(s["supervised_rows"])

    pos_rate = float(target_1 / total) if total > 0 else 0.0

    summary_table_rows.append({
        "scenario_name": s["scenario_name"],
        "K": s["k_before"],
        "H": s["horizon_windows"],
        "horizon_minutes": s["horizon_minutes"],
        "supervised_rows": total,
        "target_0": target_0,
        "target_1": target_1,
        "positive_rate": pos_rate,
        "normal_target_1": s["normal_target_1"],
        "before_target_0": s["before_target_0"],
        "before_target_1": s["before_target_1"],
        "target_equals_before_state": s["target_equals_before_state"],
        "all_positive_are_before": s["all_positive_are_before"],
        "all_before_are_positive": s["all_before_are_positive"],
    })

summary_table = pd.DataFrame(summary_table_rows)

print("\n" + "=" * 80)
print("Resumo consolidado dos cenários")
display(summary_table)

# ─────────────────────────────────────────────────────────────
# BLOCO 8 — Summaries JSON
# ─────────────────────────────────────────────────────────────

summary = {
    "notebook": "NB06",
    "purpose": "label_states_and_build_anticipation_datasets",
    "window_minutes": int(WINDOW_MINUTES),
    "features_file": str(FEATURES_FILE),
    "episodes_file": str(EPISODES_FILE),
    "features_rows": int(len(df_features)),
    "features_cols": int(df_features.shape[1]),
    "feature_bucket_min": int(bucket_min),
    "feature_bucket_max": int(bucket_max),
    "episodes_original": int(len(df_episodes)),
    "episodes_effective": int(len(effective_episodes)),
    "expected_during_effective_windows": int(expected_during_effective),
    "primary_scenario": PRIMARY_SCENARIO_NAME,
    "primary_k": int(primary_k),
    "primary_h": int(primary_h),
    "primary_horizon_minutes": int(primary_h * WINDOW_MINUTES),
    "primary_summary": primary_summary,
    "scenario_summaries": scenario_summaries,
    "scenario_artifacts": scenario_artifacts,
    "output_labeled_primary": str(OUT_LABELED_PRIMARY),
    "output_supervised_primary": str(OUT_SUPERVISED_PRIMARY),
    "output_transition_legacy_alias": str(OUT_TRANSITION_LEGACY_ALIAS),
}

sensitivity_summary = {
    "notebook": "NB06",
    "purpose": "horizon_and_state_window_sensitivity",
    "interpretation_note": (
        "The legacy K=12,H=12 configuration is retained for comparison. "
        "The revised primary scenario uses K=24,H=12 to decouple the state-labeling "
        "window from the supervised prediction horizon while preserving a 60-minute "
        "prediction horizon."
    ),
    "summary_table": summary_table.to_dict("records"),
    "scenario_summaries": scenario_summaries,
}

save_json(SUMMARY_FILE, summary)
save_json(SENSITIVITY_SUMMARY_FILE, sensitivity_summary)

log(f"Resumo principal salvo: {SUMMARY_FILE}")
log(f"Resumo de sensibilidade salvo: {SENSITIVITY_SUMMARY_FILE}")

# ─────────────────────────────────────────────────────────────
# BLOCO 9 — Resumo final na tela
# ─────────────────────────────────────────────────────────────

print("\n=== RESUMO FINAL — NB06 ===")
print(f"Features: {df_features.shape}")
print(f"Episódios originais: {len(df_episodes)}")
print(f"Episódios efetivos: {len(effective_episodes)}")
print(f"Janelas DURING efetivas: {expected_during_effective}")
print()
print(f"Cenário principal: {PRIMARY_SCENARIO_NAME}")
print(f"K principal: {primary_k}")
print(f"H principal: {primary_h} janelas ({primary_h * WINDOW_MINUTES} minutos)")
print()

print("Resumo do cenário principal:")
display(pd.DataFrame([{
    "scenario": primary_summary["scenario_name"],
    "K": primary_summary["k_before"],
    "H": primary_summary["horizon_windows"],
    "supervised_rows": primary_summary["supervised_rows"],
    "state_distribution": primary_summary["state_distribution"],
    "target_distribution": primary_summary["target_distribution"],
    "normal_target_1": primary_summary["normal_target_1"],
    "before_target_0": primary_summary["before_target_0"],
    "before_target_1": primary_summary["before_target_1"],
    "target_equals_before_state": primary_summary["target_equals_before_state"],
}]))

print()
print("Resumo consolidado dos cenários:")
display(summary_table)

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull)...
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH = /content/drive/MyDrive/Mestrado/04-reports
[NB06_labeling_states] Features carregadas: shape=(8927, 28)
[NB06_labeling_states] Episódios carregados: shape=(285, 18)
[NB06_labeling_states] Faixa de bucket_id nas features: 3..8929
[NB06_labeling_states] Episódios originais: 285
[NB06_labeling_states] Episódios efetivos na base de features: 284
[NB06_labeling_states] Janelas DURING efetivas esperadas: 389
[NB06_labeling_states] Executando cenário: legacy_K12_H12 (K=12, H=12)

Cenário: legacy_K12_H12
K=12 | H=12
Distribuição de estados:


,count
AFTER,979
BEFORE,1987
DURING,389
NORMAL,5572


Distribuição do alvo:


,count
0,5572
1,1987


Tabela state × target:


transition_target,0,1
state,,
BEFORE,0,1987
NORMAL,5572,0


Diagnóstico de equivalência:
target_equals_before_state: True
all_positive_are_before: True
all_before_are_positive: True
NORMAL com target=1: 0
BEFORE com target=0: 0
BEFORE com target=1: 1987
[NB06_labeling_states] Executando cenário: sensitivity_K12_H06 (K=12, H=6)

Cenário: sensitivity_K12_H06
K=12 | H=6
Distribuição de estados:


,count
AFTER,979
BEFORE,1987
DURING,389
NORMAL,5572


Distribuição do alvo:


,count
0,6303
1,1256


Tabela state × target:


transition_target,0,1
state,,
BEFORE,731,1256
NORMAL,5572,0


Diagnóstico de equivalência:
target_equals_before_state: False
all_positive_are_before: True
all_before_are_positive: False
NORMAL com target=1: 0
BEFORE com target=0: 731
BEFORE com target=1: 1256
[NB06_labeling_states] Executando cenário: sensitivity_K12_H03 (K=12, H=3)

Cenário: sensitivity_K12_H03
K=12 | H=3
Distribuição de estados:


,count
AFTER,979
BEFORE,1987
DURING,389
NORMAL,5572


Distribuição do alvo:


,count
0,6806
1,753


Tabela state × target:


transition_target,0,1
state,,
BEFORE,1234,753
NORMAL,5572,0


Diagnóstico de equivalência:
target_equals_before_state: False
all_positive_are_before: True
all_before_are_positive: False
NORMAL com target=1: 0
BEFORE com target=0: 1234
BEFORE com target=1: 753
[NB06_labeling_states] Executando cenário: primary_K24_H12 (K=24, H=12)

Cenário: primary_K24_H12
K=24 | H=12
Distribuição de estados:


,count
AFTER,1271
BEFORE,2966
DURING,389
NORMAL,4301


Distribuição do alvo:


,count
0,5280
1,1987


Tabela state × target:


transition_target,0,1
state,,
BEFORE,979,1987
NORMAL,4301,0


Diagnóstico de equivalência:
target_equals_before_state: False
all_positive_are_before: True
all_before_are_positive: False
NORMAL com target=1: 0
BEFORE com target=0: 979
BEFORE com target=1: 1987
[NB06_labeling_states] Executando cenário: sensitivity_K24_H06 (K=24, H=6)

Cenário: sensitivity_K24_H06
K=24 | H=6
Distribuição de estados:


,count
AFTER,1271
BEFORE,2966
DURING,389
NORMAL,4301


Distribuição do alvo:


,count
0,6011
1,1256


Tabela state × target:


transition_target,0,1
state,,
BEFORE,1710,1256
NORMAL,4301,0


Diagnóstico de equivalência:
target_equals_before_state: False
all_positive_are_before: True
all_before_are_positive: False
NORMAL com target=1: 0
BEFORE com target=0: 1710
BEFORE com target=1: 1256
[NB06_labeling_states] Executando cenário: sensitivity_K24_H03 (K=24, H=3)

Cenário: sensitivity_K24_H03
K=24 | H=3
Distribuição de estados:


,count
AFTER,1271
BEFORE,2966
DURING,389
NORMAL,4301


Distribuição do alvo:


,count
0,6514
1,753


Tabela state × target:


transition_target,0,1
state,,
BEFORE,2213,753
NORMAL,4301,0


Diagnóstico de equivalência:
target_equals_before_state: False
all_positive_are_before: True
all_before_are_positive: False
NORMAL com target=1: 0
BEFORE com target=0: 2213
BEFORE com target=1: 753
[NB06_labeling_states] Base rotulada principal salva: /content/drive/MyDrive/Mestrado/02-datasets/03-features/window_5min_labeled.parquet
[NB06_labeling_states] Base supervisionada principal salva: /content/drive/MyDrive/Mestrado/02-datasets/03-features/anticipation_supervised_dataset.parquet
[NB06_labeling_states] Alias legado salvo: /content/drive/MyDrive/Mestrado/02-datasets/03-features/transition_dataset.parquet

Resumo consolidado dos cenários


,scenario_name,K,H,horizon_minutes,supervised_rows,target_0,target_1,positive_rate,normal_target_1,before_target_0,before_target_1,target_equals_before_state,all_positive_are_before,all_before_are_positive
0,legacy_K12_H12,12,12,60,7559,5572,1987,0.262865,0,0,1987,True,True,True
1,sensitivity_K12_H06,12,6,30,7559,6303,1256,0.166160,0,731,1256,False,True,False
2,sensitivity_K12_H03,12,3,15,7559,6806,753,0.099616,0,1234,753,False,True,False
3,primary_K24_H12,24,12,60,7267,5280,1987,0.273428,0,979,1987,False,True,False
4,sensitivity_K24_H06,24,6,30,7267,6011,1256,0.172836,0,1710,1256,False,True,False
5,sensitivity_K24_H03,24,3,15,7267,6514,753,0.103619,0,2213,753,False,True,False


[NB06_labeling_states] Resumo principal salvo: /content/drive/MyDrive/Mestrado/04-reports/06_labeling_states_summary.json
[NB06_labeling_states] Resumo de sensibilidade salvo: /content/drive/MyDrive/Mestrado/04-reports/06_horizon_sensitivity_summary.json

=== RESUMO FINAL — NB06 ===
Features: (8927, 28)
Episódios originais: 285
Episódios efetivos: 284
Janelas DURING efetivas: 389

Cenário principal: primary_K24_H12
K principal: 24
H principal: 12 janelas (60 minutos)

Resumo do cenário principal:


,scenario,K,H,supervised_rows,state_distribution,target_distribution,normal_target_1,before_target_0,before_target_1,target_equals_before_state
0,primary_K24_H12,24,12,7267,"{'AFTER': 1271, 'BEFORE': 2966, 'DURING': 389,...","{'0': 5280, '1': 1987}",0,979,1987,False



Resumo consolidado dos cenários:


,scenario_name,K,H,horizon_minutes,supervised_rows,target_0,target_1,positive_rate,normal_target_1,before_target_0,before_target_1,target_equals_before_state,all_positive_are_before,all_before_are_positive
0,legacy_K12_H12,12,12,60,7559,5572,1987,0.262865,0,0,1987,True,True,True
1,sensitivity_K12_H06,12,6,30,7559,6303,1256,0.166160,0,731,1256,False,True,False
2,sensitivity_K12_H03,12,3,15,7559,6806,753,0.099616,0,1234,753,False,True,False
3,primary_K24_H12,24,12,60,7267,5280,1987,0.273428,0,979,1987,False,True,False
4,sensitivity_K24_H06,24,6,30,7267,6011,1256,0.172836,0,1710,1256,False,True,False
5,sensitivity_K24_H03,24,3,15,7267,6514,753,0.103619,0,2213,753,False,True,False


# Conclusão da Derivação de Estados e Formulação Supervisionada

A etapa de derivação de estados e formulação supervisionada foi concluída com sucesso.

Nesta versão, o notebook separou explicitamente dois parâmetros que antes estavam acoplados:

1. `K`: janela usada para definir os estados `BEFORE` e `AFTER`;
2. `H`: horizonte futuro usado para definir o alvo supervisionado de antecipação.

A versão anterior utilizava `K = 12` e `H = 12`, o que produzia equivalência direta entre o estado `BEFORE` e a classe positiva da tarefa supervisionada. Esta execução manteve esse cenário legado apenas como referência comparativa e adotou como cenário principal revisado:

| Parâmetro | Valor |
|---|---:|
| `K` | 24 janelas |
| `H` | 12 janelas |
| Horizonte preditivo | 60 minutos |
| Janela pré-crítica | 120 minutos |

Com isso, o NB06 passou a produzir uma base supervisionada principal em que o estado `BEFORE` não é mais equivalente ao alvo positivo.

## 12. Resultados Gerais da Execução

A base de features utilizada nesta etapa apresentou:

| Item | Valor |
|---|---:|
| Linhas na base de features | 8.927 |
| Colunas na base de features | 28 |
| Faixa de `bucket_id` | 3..8.929 |
| Episódios críticos originais | 285 |
| Episódios efetivos dentro da faixa das features | 284 |
| Janelas efetivas em estado `DURING` | 389 |

A diferença entre episódios originais e episódios efetivos decorre do recorte inicial da base de features, que começa em `bucket_id = 3` após a remoção mínima necessária para construção dos lags no NB05.

## 13. Cenário Legado: K=12, H=12

No cenário legado, foram obtidos:

| Métrica | Valor |
|---|---:|
| `supervised_rows` | 7.559 |
| `target_0` | 5.572 |
| `target_1` | 1.987 |
| `positive_rate` | 0,2629 |
| `before_target_0` | 0 |
| `before_target_1` | 1.987 |
| `normal_target_1` | 0 |
| `target_equals_before_state` | `true` |

A tabela cruzada `state × transition_target` mostrou que:

1. todas as janelas `NORMAL` receberam alvo 0;
2. todas as janelas `BEFORE` receberam alvo 1.

Esse resultado confirma a crítica metodológica: quando `K = H = 12`, a classe positiva coincide com o estado `BEFORE`. Assim, esse cenário não deve ser tratado como formulação principal, mas apenas como cenário legado de comparação.

## 14. Cenário Principal Revisado: K=24, H=12

No cenário principal revisado, foram obtidos:

| Métrica | Valor |
|---|---:|
| `supervised_rows` | 7.267 |
| `target_0` | 5.280 |
| `target_1` | 1.987 |
| `positive_rate` | 0,2734 |
| `before_target_0` | 979 |
| `before_target_1` | 1.987 |
| `normal_target_1` | 0 |
| `target_equals_before_state` | `false` |

A distribuição dos estados no cenário principal foi:

| Estado | Janelas |
|---|---:|
| `NORMAL` | 4.301 |
| `BEFORE` | 2.966 |
| `DURING` | 389 |
| `AFTER` | 1.271 |

A tabela cruzada `state × transition_target` indicou:

| Estado | Alvo 0 | Alvo 1 |
|---|---:|---:|
| `NORMAL` | 4.301 | 0 |
| `BEFORE` | 979 | 1.987 |

Esse resultado mostra que o alvo positivo deixou de ser equivalente ao estado `BEFORE`, pois uma parte relevante das janelas `BEFORE` recebeu alvo 0.

A tarefa supervisionada passou a diferenciar, dentro da região pré-crítica ampliada, janelas que entram em `DURING` dentro do horizonte futuro de 60 minutos daquelas que ainda não entram nesse horizonte.

## 15. Cenários de Sensibilidade

Além do cenário principal, foram avaliados cenários com horizontes menores.

Para `K = 12`:

| Cenário | H | `target_0` | `target_1` | `before_target_0` | `before_target_1` | `target_equals_before_state` |
|---|---:|---:|---:|---:|---:|---|
| `sensitivity_K12_H06` | 6 | 6.303 | 1.256 | 731 | 1.256 | `false` |
| `sensitivity_K12_H03` | 3 | 6.806 | 753 | 1.234 | 753 | `false` |

Para `K = 24`:

| Cenário | H | `target_0` | `target_1` | `before_target_0` | `before_target_1` | `target_equals_before_state` |
|---|---:|---:|---:|---:|---:|---|
| `sensitivity_K24_H06` | 6 | 6.011 | 1.256 | 1.710 | 1.256 | `false` |
| `sensitivity_K24_H03` | 3 | 6.514 | 753 | 2.213 | 753 | `false` |

Esses resultados indicam que a redução do horizonte `H` diminui a quantidade de exemplos positivos e aumenta a quantidade de janelas `BEFORE` com alvo 0, reforçando a sensibilidade da formulação supervisionada à escolha do horizonte temporal.

## 16. Diagnóstico de Circularidade

O diagnóstico de circularidade mostrou que:

| Cenário | `target_equals_before_state` |
|---|---|
| `legacy_K12_H12` | `true` |
| `sensitivity_K12_H06` | `false` |
| `sensitivity_K12_H03` | `false` |
| `primary_K24_H12` | `false` |
| `sensitivity_K24_H06` | `false` |
| `sensitivity_K24_H03` | `false` |

Portanto, a revisão do NB06 mitigou a circularidade mais forte da formulação anterior.

Entretanto, no cenário principal, todos os casos positivos continuam pertencendo ao estado `BEFORE`. Assim, a formulação deve ser interpretada como antecipação dentro de uma região pré-crítica ampliada, e não como predição totalmente independente da definição dos estados operacionais.

## 17. Importância para o Pipeline Analítico

O NB06 revisado melhora a validade metodológica do pipeline por três razões principais.

Primeiro, documenta explicitamente o problema do cenário legado, em que `K = H = 12` fazia a classe positiva coincidir com o estado `BEFORE`.

Segundo, adota como cenário principal uma configuração desacoplada, com `K = 24` e `H = 12`, preservando o horizonte preditivo de 60 minutos e ampliando a janela pré-crítica para 120 minutos.

Terceiro, gera cenários de sensibilidade que permitem avaliar o efeito de diferentes horizontes de antecipação sobre a distribuição do alvo supervisionado.

Essa alteração torna a modelagem posterior mais defensável, pois o modelo não será treinado apenas para distinguir `NORMAL` de `BEFORE`. No cenário principal, ele também deverá distinguir, dentro do próprio estado `BEFORE`, janelas que entrarão em `DURING` dentro do horizonte de 60 minutos daquelas que ainda não entrarão nesse horizonte.

## 18. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos principais:

| Artefato | Finalidade |
|---|---|
| `window_5min_labeled.parquet` | Base rotulada principal |
| `anticipation_supervised_dataset.parquet` | Base supervisionada principal |
| `transition_dataset.parquet` | Alias legado para compatibilidade |
| `06_labeling_states_summary.json` | Summary principal |
| `06_horizon_sensitivity_summary.json` | Summary de sensibilidade |

Também foram gerados artefatos por cenário:

| Artefato | Finalidade |
|---|---|
| `window_5min_labeled_K12.parquet` | Base rotulada com K=12 |
| `window_5min_labeled_K24.parquet` | Base rotulada com K=24 |
| `anticipation_supervised_dataset_K12_H12.parquet` | Base supervisionada do cenário legado |
| `anticipation_supervised_dataset_K12_H06.parquet` | Sensibilidade K=12, H=6 |
| `anticipation_supervised_dataset_K12_H03.parquet` | Sensibilidade K=12, H=3 |
| `anticipation_supervised_dataset_K24_H12.parquet` | Cenário principal revisado |
| `anticipation_supervised_dataset_K24_H06.parquet` | Sensibilidade K=24, H=6 |
| `anticipation_supervised_dataset_K24_H03.parquet` | Sensibilidade K=24, H=3 |

O arquivo canônico `anticipation_supervised_dataset.parquet` passa a representar o cenário principal revisado `primary_K24_H12`.

O arquivo `transition_dataset.parquet` foi mantido apenas como alias legado para compatibilidade com notebooks posteriores.

## 19. Encaminhamento do Pipeline

A partir desta execução, as etapas posteriores devem utilizar como base supervisionada principal:

```text
anticipation_supervised_dataset.parquet